src/plugins/typedispatcher/helpers.c

```diff
@@ -0,0 +1,437 @@
+	Key * key;
+	ksRewind (types);
+	while ((key = ksNext (types)) != NULL)
+	{
+		TypeConfig * tc = *(TypeConfig **)keyValue (key);
+		ksDel (tc->types);
+		ksDel (tc->params);
+		keyDel (tc->scope);
+		freeType (tc);
+	}
+}
+
+// helper for closeDispatch
+// close all loaded plugins
+static void closePlugins (KeySet * plugins)
+{
+	if (!plugins) return;
+	Key * key;
+	ksRewind (plugins);
+	while ((key = ksNext (plugins)) != NULL)
+	{
+		PluginConfig * pc = *(PluginConfig **)keyValue (key);
+		if (pc->plugin)
+		{
+			elektraPluginClose (pc->plugin, NULL);
+		}
+		elektraFree (pc);
+	}
+}
+
+
+// close and free everything
+void closeDispatchConfig (Plugin * handle)
+{
+	DispatchConfig * config = elektraPluginGetData (handle);
+	if (!config)
+	{
+		return;
+	}
+	closePlugins (config->plugins);
+	ksDel (config->plugins);
+	elektraModulesClose (config->modules, NULL);
+	ksDel (config->modules);
+	freeTypes (config->types);
+	ksDel (config->types);
+	elektraFree (config);
+}
+
+
+// initialize plugin configuration
+DispatchConfig * initDispatchConfig ()
+{
+	DispatchConfig * config = NULL;
+	config = elektraCalloc (sizeof (DispatchConfig));
+	if (!config) return NULL;
+	config->onError = FAIL;
+	config->plugins = ksNew (0, KS_END);
+	config->modules = ksNew (0, KS_END);
+	config->types = ksNew (0, KS_END);
+	elektraModulesInit (config->modules, NULL);
+	return config;
+}
+
+static KeySet * getKeysBelow (const Key * key, KeySet * ks, KeyRelType rel)
+{
+	if (!key) return NULL;
+	if (!ks) return NULL;
+	ksRewind (ks);
+	KeySet * result = NULL;
+	result = ksNew (ksGetSize (ks), KS_END);
+	Key * cur;
+	while ((cur = ksNext (ks)) != NULL)
+	{
+		int r = keyRel2 (key, cur, rel);
+		if (r > 0)
+		{
+#if defined(DEVBUILD) && defined(VERBOSEBUILD)
+			fprintf (stderr, "\t Key %s is %d levels below %s\n", keyName (cur), r, keyName (key));
+#endif
+			ksAppendKey (result, cur);
+		}
+	}
+	ksRewind (result);
+	return result;
+}
+
+// return a keyset containing all keys below <key>
+KeySet * getAllKeysBelow (const Key * key, KeySet * ks)
+{
+	return getKeysBelow (key, ks, ELEKTRA_REL_BELOW_SAME_NS);
+}
+
+// return a keyset containing all keys directly below <key>
+KeySet * getKeysDirectBelow (const Key * key, KeySet * ks)
+{
+	return getKeysBelow (key, ks, ELEKTRA_REL_DIRECT_BELOW_SAME_NS);
+}
+
+// check if key is withing scope of datatype
+// key is within scope if either the key is in the same namespace
+// and below the key where the datatype has been defined, or if
+// key is below the key where the datatype has been defined and the
+// datatype has been defined in the spec namespace
+
+int isWithinScope (const TypeConfig * config, const Key * key)
+{
+	const Key * scope = config->scope;
+	if (keyCmp (scope, key))
+	{
+		if (keyRel2 (scope, key, ELEKTRA_REL_BELOW_SAME_NS) <= 0)
+		{
+			elektraNamespace scopeNS = keyGetNamespace (scope);
+			if (scopeNS == KEY_NS_SPEC)
+			{
+				if (keyRel2 (scope, key, ELEKTRA_REL_BELOW_IGNORE_NS) <= 0)
```

Comment:
in other parts of the code you don't use braces for single line return statements though cause of the nested structure here it might be appropriate. Maybe even simplify this as
return keyRel2 (scope, key, ELEKTRA_REL_BELOW_IGNORE_NS) > 0; ?